### DISCUSSION TITLE
Is it still possible to declare Head elements client side when using the App Directory ?

### DISCUSSION THREAD

**Discussion by swve** [vercel/next.js#49704] [App Router]
Hi all, 

Now that `next/head` isn't supported in the App Directory and the `metadata` object is only supported in Server Components does it mean we cannot declare `Head` elements in client components anymore ? 

Also the `head.js` file that was introduced isn't in the docs anymore https://nextjs.org/docs/app

**Comment by volpster2100** [comment]
You can! If you need dynamic head data, you can re-use a fetch call since they’re de-duped to populate page-specific meta

https://nextjs.org/docs/app/building-your-application/optimizing/metadata#seo

**Comment by volpster2100** [comment]
I responded too quickly. You can use it in a page or layout and a page can be a client component.

Just curious what need it’s filling to use it in client components specifically?

**Reply by swve** [reply]
You can use `next/head` only in the Pages directory, there is no support for the App Directory (tried on 13.3) 

In migration scenarios sometimes you would want to just add `use client;` and have your previously working `Pages Directory` files in the `App Directory` without much reformat (having a server comp and a client comp), I guess Clients component are limited now and more and more features are going to be exclusive to Server Components.

Now let's say you have 30 client comp pages, you'll have to reformat all of them one-shot just because you want your pages to have a title.

**Reply by swve** [reply]
Server Components are great and I love all the possibilities it's giving but I just think having retro features like `next/head` in Client Components make it easier for gradual migrations, especially big projects.

**Comment by nick-kang** [comment]
Not the cleanest solution, but I was able to get the <meta/> tags working on 'use client'. This only works if you don't need to persist Context state as you navigate your app.

You pretty much have to manually recreate the entire <html/> tree every page. The drawback is that you can't persist state in your Context Providers.

```typescript
// app/layout.tsx
export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}): React.ReactNode {
  return children
}
```

```typescript
// app/page.tsx
'use client'

export default function Page(): JSX.Element {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta content="width=device-width" name="viewport" />
        <meta content="Root page" name="title" />
      </head>
      <body>
        // your content
      </body>
    </html>
}
```

```typescript
// app/another-page/page.tsx
'use client'

export default function Page(): JSX.Element {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta content="width=device-width" name="viewport" />
        <meta content="Subpage" name="title" />
      </head>
      <body>
        // your content
      </body>
    </html>
}
```

**Comment by martinjnilsen** [comment]
I do think I found a better fitting solution for my own project, and I wanted to share in case anyone else would benefit from it.

**Problem**
The problem that I had, was that I was unable to properly set my meta theme color with the use of [generateViewport](https://nextjs.org/docs/app/api-reference/functions/generate-viewport). This is due to the usage of MUI as a styling library, and the theme I have is thus only available client side. If the themes only depended on the preferred color scheme, the generateViewport function would suffice, but as I have a color toggle capability on the site, some client side logic would need to be applied. 

**Thoughts**
- First I set out to include the approach of `next/head` from the Pages router. This is not supported as previously stated.
- Using cookies for the theme could work, however, the approach came with some drawbacks. I would need to introduce cookies just for this case, update the cookie when toggling the theme, and the usage of cookies [would enforce](https://nextjs.org/docs/app/api-reference/functions/cookies) every page using it to be [dynamically rendered](https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-rendering). These drawbacks led me to look for another solution.
- I did also consider using [server actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations), thinking I might be able to do a function call back on the server and update the meta tag there. I am not sure if this could work, but I did not look further into the suggested approach.

**Solution**
Cue my client side function for applying the theme-color meta tag:

```tsx
"use client";
import { useTheme } from "@/styles/themes/ThemeProvider";
import { useEffect } from "react";

const ViewportMeta = () => {
	const { theme } = useTheme();

	useEffect(() => {
		const metaViewport = document.querySelector('meta[name="theme-color"]');
		if (metaViewport) {
			metaViewport.setAttribute("content", theme.palette.primary.main);
		} else {
			const newMetaViewport = document.createElement("meta");
			newMetaViewport.name = "theme-color";
			newMetaViewport.content = theme.palette.primary.main;
			document.head.appendChild(newMetaViewport);
		}
	}, [theme]);

	return null;
};

export default ViewportMeta;
```

Which could be set in the `layout.tsx` file:
```diff
export default async function RootLayout({
	children,
}: {
	children: React.ReactNode;
}) {

	return (
		<html lang="en">
			<head></head>
			<body>
				<CustomThemeProvider>
+					<ViewportMeta />
					{children}
				</CustomThemeProvider>
			</body>
		</html>
	);
}
```
> This is not a full version of my layout, but simplified for the essence of this post

After some testing, this seem to yield the correct theme color, while also responding to theme toggling.
